### ISSUE TITLE
Prism use the same depth for variables (just) outside and inside the eval

### ISSUE THREAD

**Issue by eregon** [ruby/prism#2192]
```
$ ruby -Ilib -rprism -e 'pp Prism.parse("b; c = 3", scopes: [[:b]])'      
#<Prism::ParseResult:0x00007fa688ccef80
 @comments=[],
 @data_loc=nil,
 @errors=[],
 @magic_comments=[],
 @source=#<Prism::Source:0x00007fa688c986b0 @offsets=[0], @source="b; c = 3", @start_line=1>,
 @value=
  @ ProgramNode (location: (1,0)-(1,8))
  ├── locals: [:b, :c]
  └── statements:
      @ StatementsNode (location: (1,0)-(1,8))
      └── body: (length: 2)
          ├── @ LocalVariableReadNode (location: (1,0)-(1,1))
          │   ├── name: :b
          │   └── depth: 0
          └── @ LocalVariableWriteNode (location: (1,3)-(1,8))
              ├── name: :c
              ├── depth: 0
              ├── name_loc: (1,3)-(1,4) = "c"
              ├── value:
              │   @ IntegerNode (location: (1,7)-(1,8))
              │   └── flags: decimal
              └── operator_loc: (1,5)-(1,6) = "=",
 @warnings=[]>
```

So this output shows that `b` and `c` would be in the same scope.
But they are not:
```
$ ruby -e 'b=2; eval "c=3"; p binding.local_variables'              
[:b]
```
So I think the depth for all variables in `scopes:` should be incremented by 1.

**Comment by kddnewton** [comment]
This was the way it was originally designed, but we changed this when we had some spec failures in @enebo's test run. Sorry Tom I can't remember what the original issue was.

**Comment by enebo** [comment]
@kddnewton @eregon The problem was supporting bindings. If I pass in a binding I need the ProgramNode scope to include those variables into the same scope as the binding scope.  Otherwise it is a new scope on top of the scopes passed in. 

I seem to pass all specs and tests I run at it but I realize how prism works now it assumes any call the first scope is a binding or a brand new empty scope.

Even though I made this work for me I guess it may be a lot clearer as an API to specify whether you are passing in a binding or not.  I may just keep this 100% binding eval in my codebase since the legacy parser passes in the extra scope to the parser but I can see how this is confusing.

**Comment by eregon** [comment]
Interesting, indeed with a Binding I think there is no way to differentiate these two scopes, the one accessed by `Binding#local_variable_set` and the one used in `Binding#eval`.
There might be separate scopes in a Ruby implementation (notably the case on TruffleRuby) but I did not find a way to reveal it at the Ruby level:
```
$ ruby -e 'def new_binding; binding; end; b=new_binding; b.local_variable_set(:x, 1); b.eval("y = 2"); p b.local_variables'
[:y, :x]

Also affects Kernel#eval when passed a binding:
$ ruby -e 'def new_binding; binding; end; b=new_binding; b.local_variable_set(:x, 1); eval("y = 2", b); p b.local_variables'
[:y, :x]
```

I think this is fine then but it would be great to document it as it is not obvious.

And indeed maybe a boolean option for whether variables in the code to parse get a new scope or reuse the scope of the first of the `scopes:` passed. But it might be simpler to just pass an extra empty scope if that's documented.

**Comment by enebo** [comment]
> And indeed maybe a boolean option for whether variables in the code to parse get a new scope or reuse the scope of the first of the `scopes:` passed. But it might be simpler to just pass an extra empty scope if that's documented (or use a depth offset since we'll anyway need one for `for` in truffleruby).

Yeah this is exactly the two options at play and we effectively do the second one now.  I think having the first option just add 'binding = true' means you will assume first scope passed in is the zeroth scope would probably lead to less confusion but I would probably just always pass in that and then add an extra scope (since this is how it works with the legacy parser and I still support both).

In other words, I am ok going either direction but I want to do what causes the least amount of confusion.

**Comment by eregon** [comment]
I am thinking there are 3 "cases" for parsing:
* parsing a file
* parsing an eval without a binding
* parsing an eval with a binding

For the first two they need their own scope, for the 3rd it depends on the Ruby implementation but semantically a new scope is not required.
We could also see the 2nd one as creating a new Binding since none was passed, and then it's equivalent to the 3rd one (which how we deal with it in Kernel#eval in TruffleRuby).
If we see it like that, then it's just 2 cases, eval vs non-eval, and that's I think already known via `scopes.length == 0 ? "file" : "eval"` because an eval always has at least 1 scope around it (I think).

So I think documenting is the best here, and a new API (like binding: true/false or new_scope: true/false) is probably confusing/error-prone depending on the POV on these 3 cases.

**Comment by kddnewton** [comment]
I think we're good on the documentation here with the scopes option now. If it's insufficient we can revisit.
